test/pylib/cpp/boost/prepare_instance.py

```diff
@@ -0,0 +1,85 @@
+from test.pylib.minio_server import MinioServer
+from test.pylib.s3_server_mock import MockS3Server
+from test.pylib.util import LogPrefixAdapter
+
+
+class PrepareWorker:
+    def __init__(self, env_file: Path):
+        self.env_file = env_file
+
+    def prepare(self) -> None:
+        timeout = 10
+        start_time = time.time()
+        while True:
+            if os.path.exists(self.env_file):
+                with open(self.env_file, 'r') as file:
+                    for line in file.readlines():
+                        key, value = line.strip().split('=', 1)
+                        os.environ[key] = value
+                break
+
+            if time.time() - start_time > timeout:
+                raise TimeoutError(f"Timeout waiting for file {self.env_file}")
+            # Sleep needed to wait when the controller will create a file with environment variables.
+            # Without sleep checking of the file existence will be too fast,
+            # so it will finish before the file is created
+            time.sleep(0.5)
+
+    def cleanup(self) -> None:
+        pass
+
+
+class PrepareController:
+    def __init__(self, temp_dir: Path, modes: list[str], env_file: Path):
+        self.temp_dir = temp_dir
+        pytest_dirs = [self.temp_dir / mode / 'pytest' for mode in modes]
+        for directory in [self.temp_dir, *pytest_dirs]:
+            if not directory.exists():
+                directory.mkdir()
+        self.env_file = env_file
+        hosts = HostRegistry()
+        self.loop = asyncio.new_event_loop()
+        address = self.loop.run_until_complete(hosts.lease_host())
+        self.ms = MinioServer(self.temp_dir, '127.0.0.1', LogPrefixAdapter(logging.getLogger('minio'), {'prefix': 'minio'}))
+        self.mock_s3 = MockS3Server(address, 2012,
+                                    LogPrefixAdapter(logging.getLogger('s3_mock'), {'prefix': 's3_mock'}))
+
+    def prepare(self) -> None:
+        tasks = [self.loop.create_task(self.ms.start()), self.loop.create_task(self.mock_s3.start())]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
+        envs = self.ms.get_envs_settings()
+        envs.update(self.mock_s3.get_envs_settings())
+        with open(self.env_file, 'w') as file:
+            for key, value in envs.items():
+                file.write(f"{key}={value}\n")
+
+    def cleanup(self) -> None:
+        tasks = [self.loop.create_task(self.ms.stop()), self.loop.create_task(self.mock_s3.stop())]
+        self.loop.run_until_complete(asyncio.gather(*tasks))
+        if os.path.exists(self.env_file):
+            self.env_file.unlink()
+
+def get_prepare_instance(temp_dir: Path, is_worker: bool, modes: list[str]) -> PrepareWorker | PrepareController:
+    """
+    With xdist use there's a need to pass environment variables from controller to workers. According classes responsible
+    for configuration controller and workers. But start servers will be triggered only once on the controller.
```

Comment:
When xdist is enabled, we have two types of processes: controller and workers. Actually, we need to distinguish only the workers, since they have their own logic. That is why when xdist is disabled code for controller, but in this case main, will be executed.